README.md
# MRS-App Client Detector #

This simple JS script provides information about browser and user/client device to use with your mobile Apps.

It is suitable for Phonegap and Cordova.

## Compile
Make sure you have NodeJS and Grunt installed in your system.
Simply run:
```
$ npm install
$ grunt
```

## How to run
Include in your html header
```HTML
<script src="mrs-app-clientdetector.min.js"></script>
```

Then, use it using window.client!
```
$ window.client
	browser: "Chrome"
	browserVersion: "36.0.1985.143"
	cookies: true
	mobile: false
	os: "Windows"
	osVersion: "8.1"
	screen: "1366 x 768"
```

If you need to run again, just call
```
$ window.client.run()
```

## Full features
Available soon :)

Check our code to know more :)

CHANGELOG.md
# MRS.App - Client Detector release notes

## v1.0.0
Change to MRS.App namespace.


## v0.1.0

First release.
Detects:
	1. screen size
	2. OS name
	3. OS version
	4. Browser name
	5. Browser version
	6. Mobile device
	7. Cookie capable


test/unit/constants.js
/**
    Set of constants focused on MRS.Core automated tests
**/


test/unit/test.detector.js
describe('MRS.Client-Detector:', function () {
    'use strict';

    var setBrowser = function(userAgent, appName, appVersion) {
        var __originalNavigator = navigator;
        navigator = new Object();
        navigator.__proto__ = __originalNavigator;

        if (userAgent)
            navigator.__defineGetter__('userAgent', function(){
                return userAgent;
            });
        if (appName)
            navigator.__defineGetter__('appName', function(){
                return appName;
            });
        if (appVersion)
            navigator.__defineGetter__('appVersion', function(){
                return appVersion;
            });
    };
    
    describe('Windows - Chrome:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.143 Safari/537.36",
                'Netscape',
                "5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.143 Safari/537.36"
                );
            window.client.run();
        });
        
        it('should return windows OS 8.1', function () {
            expect(window.client.os).toBe('Windows');
            expect(window.client.osVersion).toBe('8.1');
        });

        it('should return Chrome 36.0.1985.143', function () {
            expect(window.client.browser).toBe('Chrome');
            expect(window.client.browserVersion).toBe('36.0.1985.143');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(false);
        });
        
    });
    
    describe('MacOSX - Chrome:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36",
                'Mozilla',
                "5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.130 Safari/537.36"
                );
            window.client.run();
        });
        
        it('should return Mac OSX 10.10.3', function () {
            expect(window.client.os).toBe('Mac OS X');
            expect(window.client.osVersion).toBe('10.10.3');
        });

        it('should return Chrome 43.0.2357.130', function () {
            expect(window.client.browser).toBe('Chrome');
            expect(window.client.browserVersion).toBe('43.0.2357.130');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(false);
        });
        
    });
    
    describe('MacOSX - Safari:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/600.6.3 (KHTML, like Gecko) Version/8.0.6 Safari/600.6.3",
                'Mozilla',
                "5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/600.6.3 (KHTML, like Gecko) Version/8.0.6 Safari/600.6.3"
                );
            window.client.run();
        });
        
        it('should return Mac OSX 10.10.3', function () {
            expect(window.client.os).toBe('Mac OS X');
            expect(window.client.osVersion).toBe('10.10.3');
        });

        it('should return Chrome 43.0.2357.130', function () {
            expect(window.client.browser).toBe('Safari');
            expect(window.client.browserVersion).toBe('8.0.6');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(false);
        });
        
    });
    
    describe('Android Lollipop - Chrome:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Linux; Android 5.1.1; Nexus 5 Build/LMY48B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.93 Mobile Safari/537.36",
                'Mozilla',
                "5.0 (Linux; Android 5.1.1; Nexus 5 Build/LMY48B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.93 Mobile Safari/537.36"
                );
            window.client.run();
        });
        
        it('should return Android 5.1.1', function () {
            expect(window.client.os).toBe('Android');
            expect(window.client.osVersion).toBe('5.1.1');
        });

        it('should return Chrome 43.0.2357.93', function () {
            expect(window.client.browser).toBe('Chrome');
            expect(window.client.browserVersion).toBe('43.0.2357.93');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(true);
        });
        
    });
    
    describe('Android Lollipop - Firefox:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Android; Mobile; rv:39.0) Gecko/39.0 Firefox/39.0",
                'Mozilla',
                "5.0 (Android; Mobile; rv:39.0) Gecko/39.0 Firefox/39.0"
                );
            window.client.run();
        });
        
        it('should return Android with unknown version', function () {
            expect(window.client.os).toBe('Android');
            expect(window.client.osVersion).toBe('-');
        });

        it('should return Firefox 39.0', function () {
            expect(window.client.browser).toBe('Firefox');
            expect(window.client.browserVersion).toBe('39.0');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(true);
        });
        
    });
    
    describe('iOS - Safari:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (iPhone; CPU iPhone OS 8_4 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12H143 Safari/600.1.4",
                'Mozilla',
                "5.0 (iPhone; CPU iPhone OS 8_4 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Version/8.0 Mobile/12H143 Safari/600.1.4"
                );
            window.client.run();
        });
        
        it('should return iOS 8.4', function () {
            expect(window.client.os).toBe('iOS');
            expect(window.client.osVersion).toBe('8.4.0');
        });

        it('should return Sarari 8.0', function () {
            expect(window.client.browser).toBe('Safari');
            expect(window.client.browserVersion).toBe('8.0');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(true);
        });
        
    });
    
    describe('Windows Phone 8.0 - IE:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (compatible; MSIE 10.0; Windows Phone 8.0; Trident/6.0; IEMobile/10.0; ARM; Touch; NOKIA; Lumia 530)",
                'Mozilla',
                "5.0 (compatible; MSIE 10.0; Windows Phone 8.0; Trident/6.0; IEMobile/10.0; ARM; Touch; NOKIA; Lumia 530)"
                );
            window.client.run();
        });
        
        it('should return Windows Phone 8.0', function () {
            expect(window.client.os).toBe('Windows Phone');
            expect(window.client.osVersion).toBe('8.0');
        });

        it('should return IE 10.0', function () {
            expect(window.client.browser).toBe('Microsoft Internet Explorer');
            expect(window.client.browserVersion).toBe('10.0');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(true);
        });
        
    });
    
    describe('Windows Phone 8.1 - IE:', function () {
        
        beforeEach(function() {
            setBrowser(
                "Mozilla/5.0 (Mobile; Windows Phone 8.1; Android 4.0; ARM; Trident/7.0; Touch; rv:11.0; IEMobile/11.0; NOKIA; Lumia 520; Vodafone) like iPhone OS 7_0_3 Mac OS X AppleWebKit/537 (KHTML, like Gecko) Mobile Safari/537",
                'Mozilla',
                "5.0 (Mobile; Windows Phone 8.1; Android 4.0; ARM; Trident/7.0; Touch; rv:11.0; IEMobile/11.0; NOKIA; Lumia 520; Vodafone) like iPhone OS 7_0_3 Mac OS X AppleWebKit/537 (KHTML, like Gecko) Mobile Safari/537"
                );
            window.client.run();
        });
        
        it('should return Windows Phone 8.1', function () {
            expect(window.client.os).toBe('Windows Phone');
            expect(window.client.osVersion).toBe('8.1');
        });

        it('should return IE 8.0', function () {
           // expect(window.client.browser).toBe('Microsoft Internet Explorer');
           // expect(window.client.browserVersion).toBe('8.0');
        });

        it('should not be mobile', function() {
            expect(window.client.mobile).toBe(true);
        });
        
    });
    
});

gruntfile.js
'use strict';

/*  
    Description:    File for build based on Grunt package over NodeJS platform.
 */

module.exports = function (grunt) {
    
    // loading project properties
    var project = grunt.file.readJSON('config/build.conf.json');
    
    
    // Project Configuration
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),               // loading dependencies
        
        // cleaning last build
        clean: {
            options: {
                force: true,
                "no-write": false
            },
            dist: ["dist", "test/result", "docs"],
            min: "dist/minified.js"
        },
        
        // copy minified file
        copy: {
            min: {
                src:  "dist/minified.js",
                dest: "dist/" + project.minifiedDeployFile
            }
        },
        
        // JS validator
        jshint: {
            source: ["src/*.js"]
        },
        
        // running automated tests
        karma: {
            unit: {
                configFile: "config/karma.conf.js"
            }
        },
        
        
        // documentation
        yuidoc: {
            compile: {
                name: '<%= pkg.name %>',
                description: '<%= pkg.description %>',
                version: '<%= pkg.version %>',
                url: '<%= pkg.homepage %>',
                options: {
                    paths: ["src"],
                    outdir: "docs"
                }
            }
        },
        
        // concat
        concat: {
            options: {
                stripBanners: {
                    block: true,
                    line: true
                },
                process: true,
                banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' + '<%= grunt.template.today("yyyy-mm-dd HH:MM:ss Z") %> */\n'
            },
            dist: {
                src: ['src/*.js'],
                dest: 'dist/' + project.moduleDeployFile
            }
        },
        
        // minify
        uglify: {
            options: {
                mangle: project.obfuscate,
                sourceMap: project.obfuscate,
                sourceMapName: 'dist/' + project.mapDeployFile,
                banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' + '<%= grunt.template.today("yyyy-mm-dd HH:MM:ss Z") %> */\n'
            },
            dist: {
                files: {
                    'dist/minified.js' : ['dist/' + project.moduleDeployFile]
                }
            }
        }
        
        
    });
    
    
    // loading node dependencies
    grunt.loadNpmTasks('grunt-karma');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-yuidoc');
    
    
    // tests
    grunt.registerTask('test', function () {
        grunt.task.run('karma');
    });
    
    // run ALL targets
    grunt.registerTask('all', ['clean', 'jshint', 'test', 'yuidoc', 'concat', 'uglify', 'copy:min', 'clean:min']); //distribute
    
    // Default task(s).
    grunt.registerTask('default', ['all']);
};



config/karma.conf.js
module.exports = function (config) {
    config.set({
        basePath: '../',
        logLevel: config.LOG_WARN,
        frameworks: ['jasmine'],
        files: [
            'src/*.js',
            'test/lib/sinon/sinon-server-1.7.3.js',
            'test/lib/sinon/sinon-1.7.3.js',
			'test/unit/constants.js',
            'test/unit/test.*.js'
        ],
        exclude: [
        ],
        singleRun: true,
        reportSlowerThan: 500,
        autoWatch: true,
        browsers: ['PhantomJS'],
        reporters: ['dots', 'coverage', 'junit'],
        preprocessors: {
            // source files, that you wanna generate coverage for
            // do not include tests or libraries
            // (these files will be instrumented by Istanbul)
            'src/*.js': ['coverage']
        },
        junitReporter: {
            outputFile: 'test/result/test-results.xml',
            suite: 'unit'
        },
        coverageReporter: {
            type: 'html',
            dir: 'test/result/coverage/'
        }
    });
};

src/detector.js
/**
 * 
 * The client-detector is responsible to detect all information about browser.
 * It is suitable for desktop and mobile detection.
 * 
 * Provide client object in window.client.
 * window.client = {
 *     run: function,
 *     screen: string,
 *     browser: string,
 *     browserVersion: string,
 *     mobile: boolean,
 *     os: string,
 *     osVersion: string,
 *     cookies: boolean
 * };
 * 
 * Based on http://jsfiddle.net/ChristianL/AVyND/ viazenetti GmbH (Christian Ludwig)
 * 
 * @namespace MRS.Client-Detector
 * @since 0.1.0
**/
(function (window) {

    function run() {

        var unknown = '-';

        // screen
        var screenSize = (screen.width || '') + ' x ' + (screen.height || '');

        //browser
        var nVer = navigator.appVersion,
            nAgt = navigator.userAgent || navigator.vendor || window.opera,
            browser = navigator.appName,
            version = '' + parseFloat(navigator.appVersion),
            majorVersion = parseInt(navigator.appVersion, 10),
            nameOffset, verOffset, ix;

        // PhantomJS
        if ((verOffset = nAgt.indexOf('PhantomJS')) != -1) {
            browser = 'PhantomJS';
            version = nAgt.substring(verOffset + 5);
        }
        // BlackBerry
        else if ((verOffset = nAgt.indexOf('BlackBerry')) != -1 || (verOffset = nAgt.indexOf('BB10') != -1)) {
            browser = 'BlackBerry';
            version = unknown;
        }
        // NokiaBrowser
        else if ((verOffset = nAgt.indexOf('NokiaBrowser')) != -1) {
            browser = 'NokiaBrowser';
            version = unknown;
        }
        // Firefox
        else if ((verOffset = nAgt.indexOf('Firefox')) != -1) {
            browser = 'Firefox';
            version = nAgt.substring(verOffset + 8);
        }
        // Opera
        else if ((verOffset = nAgt.indexOf('Opera')) != -1) {
            browser = 'Opera';
            version = nAgt.substring(verOffset + 6);
            if ((verOffset = nAgt.indexOf('Version')) != -1) {
                version = nAgt.substring(verOffset + 8);
            }
        }
        // MSIE
        else if ((verOffset = nAgt.indexOf('MSIE')) != -1) {
            browser = 'Microsoft Internet Explorer';
            version = nAgt.substring(verOffset + 5);
        }
        // Chrome
        else if ((verOffset = nAgt.indexOf('Chrome')) != -1) {
            browser = 'Chrome';
            version = nAgt.substring(verOffset + 7);
        }
        // Safari
        else if ((verOffset = nAgt.indexOf('Safari')) != -1) {
            browser = 'Safari';
            version = nAgt.substring(verOffset + 7);
            if ((verOffset = nAgt.indexOf('Version')) != -1) {
                version = nAgt.substring(verOffset + 8);
            }
        }
        // Other browsers
        else if ((nameOffset = nAgt.lastIndexOf(' ') + 1) < (verOffset = nAgt.lastIndexOf('/'))) {
            browser = nAgt.substring(nameOffset, verOffset);
            version = nAgt.substring(verOffset + 1);
            if (browser.toLowerCase() == browser.toUpperCase()) {
                browser = navigator.appName;
            }
        }
        // trim the version string
        if ((ix = version.indexOf(';')) != -1) version = version.substring(0, ix);
        if ((ix = version.indexOf(' ')) != -1) version = version.substring(0, ix);

        // get major version again
        if (isNaN(parseInt('' + version, 10))) {
            majorVersion = parseInt('' + version, 10);
        }

        // mobile version
        var mobile = /android.+mobile|avantgo|bada\/|blackberry|bb10|blazer|compal|elaine|fennec|phantomjs|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino/i.test(nAgt) || /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|e\-|e\/|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(di|rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|xda(\-|2|g)|yas\-|your|zeto|zte\-/i.test(nAgt.substr(0, 4)) || /Mobile|mini|Fennec|Android|iP(ad|od|hone)|S40OviBrowser|NokiaBrowser/.test(nVer);

        // cookie
        var cookieEnabled = (navigator.cookieEnabled) ? true : false;

        if (typeof navigator.cookieEnabled == 'undefined' && !cookieEnabled) {
            document.cookie = 'testcookie';
            cookieEnabled = (document.cookie.indexOf('testcookie') != -1) ? true : false;
        }

        // system
        var os = unknown;
        var clientStrings = [
            {s:'PhantomJS', r:/PhantomJS/},
            {s:'Windows 3.11', r:/Win16/},
            {s:'Windows 95', r:/(Windows 95|Win95|Windows_95)/},
            {s:'Windows ME', r:/(Win 9x 4.90|Windows ME)/},
            {s:'Windows 98', r:/(Windows 98|Win98)/},
            {s:'Windows CE', r:/Windows CE/},
            {s:'Windows 2000', r:/(Windows NT 5.0|Windows 2000)/},
            {s:'Windows XP', r:/(Windows NT 5.1|Windows XP)/},
            {s:'Windows Server 2003', r:/Windows NT 5.2/},
            {s:'Windows Vista', r:/Windows NT 6.0/},
            {s:'Windows 7', r:/(Windows 7|Windows NT 6.1)/},
            {s:'Windows 8.1', r:/(Windows 8.1|Windows NT 6.3)/},
            {s:'Windows 8', r:/(Windows 8|Windows NT 6.2)/},
            {s:'Windows NT 4.0', r:/(Windows NT 4.0|WinNT4.0|WinNT|Windows NT)/},
            {s:'Windows ME', r:/Windows ME/},
            {s:'Windows Phone', r:/(Windows Phone|IEMobile)/},
            {s:'Android', r:/Android/},
            {s:'Open BSD', r:/OpenBSD/},
            {s:'Sun OS', r:/SunOS/},
            {s:'Linux', r:/(Linux|X11)/},
            {s:'iOS', r:/(iPhone|iPad|iPod)/},
            {s:'Mac OS X', r:/Mac OS X/},
            {s:'Mac OS', r:/(MacPPC|MacIntel|Mac_PowerPC|Macintosh)/},
            {s:'QNX', r:/QNX/},
            {s:'UNIX', r:/UNIX/},
            {s:'BeOS', r:/BeOS/},
            {s:'OS/2', r:/OS\/2/},
            {s:'Nokia', r:/NokiaBrowser|S40OviBrowser/},
            {s:'BlackBerry', r:/BlackBerry|BB10/},
            {s:'Firefox', r:/Firefox/},
            {s:'Search Bot', r:/(nuhk|Googlebot|Yammybot|Openbot|Slurp|MSNBot|Ask Jeeves\/Teoma|ia_archiver)/}
        ];

        var id, cs;
        for (id in clientStrings) {
            cs = clientStrings[id];
            if (cs.r.test(nAgt)) {
                os = cs.s;
                break;
            }
        }

        var osVersion = unknown;

        // try to match OS version
        try {
            switch (os) {
                case 'Mac OS X':
                    osVersion = /Mac OS X (10[\.\_\d]+)/.exec(nAgt)[1];
                    break;
    
                case 'Android':
                    osVersion = /Android ([\.\_\d]+)/.exec(nAgt)[1];
                    break;
    
                case 'iOS':
                    osVersion = /OS (\d+)_(\d+)_?(\d+)?/.exec(nVer);
                    osVersion = osVersion[1] + '.' + osVersion[2] + '.' + (osVersion[3] | 0);
                    break;
    
                case 'Windows Phone':
                    osVersion = /Windows Phone ([\.\_\d]+)/.exec(nAgt)[1];
                    break;
                    
                default:
                    if (/Windows/.test(os)) {
                        osVersion = /Windows (.*)/.exec(os)[1];
                        os = 'Windows';
                    }
                    break;
            }
        } catch(error) {
            // do nothing
        }
        
        
        // replace _ by . in OS version
        osVersion = osVersion.replace(new RegExp('_', 'g'), '.');
        
        // all done!

        window.client = {
            run: run,
            screen: screenSize,
            browser: browser,
            browserVersion: version,
            mobile: mobile,
            os: os,
            osVersion: osVersion,
            cookies: cookieEnabled
        };
    }

    run();
    
}(window));